src/main/java/in/twizmwaz/cardinal/module/modules/ctf/FlagObjective.java

```diff
@@ -0,0 +1,418 @@
+        recoverTime = post.getRecoverTime();
+    }
+
+    @Override
+    public void unload() {
+        HandlerList.unregisterAll(this);
+    }
+
+    @Override
+    public TeamModule getTeam() {
+        return owner;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public String getId() {
+        return id;
+    }
+
+    @Override
+    public boolean isTouched() {
+        return false;
+    }
+
+    @Override
+    public boolean isComplete() {
+        return false;
+    }
+
+    @Override
+    public boolean showOnScoreboard() {
+        return show;
+    }
+
+    @Override
+    public GameObjectiveScoreboardHandler getScoreboardHandler() {
+        return scoreboardHandler;
+    }
+
+    public String getDisplayName() {
+        return MiscUtil.convertDyeColorToChatColor(color) + getName();
+    }
+
+    public Post getPost() {
+        return post;
+    }
+
+    public List<Net> getNets() {
+        return nets;
+    }
+
+    public Player getPicker() {
+        return picker;
+    }
+
+    public void setPicker(Player picker) {
+        this.picker = picker;
+    }
+
+    public boolean isCarried() {
+        return getPicker() != null;
+    }
+
+    public int getRespawnTime() {
+        return respawnTime;
+    }
+
+    public boolean isRespawning() {
+        return respawning;
+    }
+
+    public boolean isShared() {
+        return shared;
+    }
+
+    public Block getCurrentFlagBlock() {
+        return currentFlagBlock;
+    }
+
+    public FilterModule getPickupFilter() {
+        return pickupFilter;
+    }
+
+    public FilterModule getCaptureFilter() {
+        return captureFilter;
+    }
+
+    public void respawnFlag() {
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            RegionModule spawn = getPost().getNextFlagSpawn();
+            Block block;
+            if (spawn instanceof PointRegion) {
+                block = ((PointRegion) spawn).getBlock();
+            } else {
+                block = spawn.getRandomPoint().getBlock();
+            }
+            spawnFlag(block, true);
+            FlagRespawnEvent e = new FlagRespawnEvent(this, getPost(), block);
+            Bukkit.getServer().getPluginManager().callEvent(e);
+            for (Player p : Bukkit.getOnlinePlayers()) {
+                p.sendMessage(new LocalizedChatMessage(ChatConstant.UI_FLAG_RESPAWNED, getDisplayName() + ChatColor.RESET).getMessage(p.getLocale()));
+            }
+            respawning = false;
+        }
+    }
+
+    private void spawnFlag(Block flagBlock, boolean post) {
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            if (flagBlock.getRelative(BlockFace.DOWN).getType().equals(Material.WATER)) {
+                if (dropOnWater) {
+                    flagBlock.getRelative(BlockFace.DOWN).setType(Material.PACKED_ICE);
+                } else {
+                    // TODO
```

Comment:
You need to make a "last valid block" variable, and update it on player move event if the position is valid.
in that move event is where you should check if water is a valid floor too, or if the "last valid block" should only update when having solid material 1 block under the player